### PR TITLE
Find benchmark cardinality misestimations

### PR DESCRIPTION
**Pull request by leowe**

### CODE REVIEW

File: src/benchmarklib/benchmark_sql_executor.cpp [hunk 1]
```diff
@@ -44,6 +45,10 @@ std::pair<SQLPipelineStatus, std::shared_ptr<const Table>> BenchmarkSQLExecutor:
     _visualize(pipeline);
   }
 
+  // Print cardinality estimations from lqp and actual cardinalities from pqp to csv for debugging purposes
+  //_print_debug_estimation_results(pipeline);
```
Comment (leowe) [inline src/benchmarklib/benchmark_sql_executor.cpp]: Remove comment

File: src/benchmarklib/benchmark_sql_executor.cpp [hunk 2]
```diff
@@ -44,6 +45,10 @@ std::pair<SQLPipelineStatus, std::shared_ptr<const Table>> BenchmarkSQLExecutor:
...
+  _write_cardinalities(pipeline);
```
Comment (leowe) [inline src/benchmarklib/benchmark_sql_executor.cpp]: Add flag to switch this method on and off similar to visualize

File: src/benchmarklib/benchmark_sql_executor.cpp [hunk 3]
```diff
@@ -139,6 +144,39 @@ void BenchmarkSQLExecutor::_compare_tables(const std::shared_ptr<const Table>& a
...
+  } 
+    
+  auto item_id = *_visualize_prefix;
+
+  std::ofstream output_file;
+  output_file.open ("./join_order_benchmark_cardinality_estimation.csv", std::ios_base::app);
+  output_file << item_id << "," << lqp_row_count << "," << pqp_row_count << "\n";
+  output_file.close();
+}
+
```
Comment (leowe) [inline src/benchmarklib/benchmark_sql_executor.cpp]: remove

File: src/lib/visualization/cardinality_writer.cpp [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+void CardinalityWriter::write_cardinalities(const std::vector<std::shared_ptr<AbstractLQPNode>>& lqp_roots,
+																						const std::vector<std::shared_ptr<AbstractOperator>>& pqp_roots,
+																						const std::string& benchmark_item_id){
+    _build_graph(lqp_roots);
+		_build_pqp_graph(pqp_roots);
+
+		//Assert(_lqp_nodes.size() != _pqp_nodes.size(), "LQP and PQP have differing number of nodes / differing structure. Can't match.");
+
+		// std::cout << "LQP SIZE " << _lqp_nodes.size() << std::endl;
+		// std::cout << "PQP SIZE " << _pqp_nodes.size() << std::endl;
```
Comment (leowe) [inline src/lib/visualization/cardinality_writer.cpp]: remove

File: src/lib/visualization/cardinality_writer.cpp [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+		const auto pqp_nodes_size = _pqp_nodes.size();
+		
+		const auto number_nodes = std::min(_lqp_nodes.size(), _pqp_nodes.size());
+
+		for (auto node = 0u; node < number_nodes; node++) {
+
+			const auto lqp_description = _lqp_nodes[node]->description();
+			const auto pqp_description = _pqp_nodes[node]->description();
+			// std::cout << _lqp_nodes[node]->description() << std::endl;
+			// std::cout << _pqp_nodes[node]->name() << " " << _pqp_nodes[node]->description() << std::endl;
```
Comment (leowe) [inline src/lib/visualization/cardinality_writer.cpp]: remove

File: src/lib/visualization/cardinality_writer.cpp [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+		//Assert(_lqp_nodes.size() != _pqp_nodes.size(), "LQP and PQP have differing number of nodes / differing structure. Can't match.");
+
+		// std::cout << "LQP SIZE " << _lqp_nodes.size() << std::endl;
+		// std::cout << "PQP SIZE " << _pqp_nodes.size() << std::endl;
+
+		const auto lqp_nodes_size = _lqp_nodes.size();
+		const auto pqp_nodes_size = _pqp_nodes.size();
+		
+		const auto number_nodes = std::min(_lqp_nodes.size(), _pqp_nodes.size());
```
Comment (leowe) [inline src/lib/visualization/cardinality_writer.cpp]: add comment: why are we taking the minimum + assumption that they magically match up (what they do most of the time)

File: src/lib/visualization/cardinality_writer.cpp [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+			
+			auto pqp_row_count = NAN;
+			const auto& performance_data = *_pqp_nodes[node]->performance_data;
+
+  		if (_pqp_nodes[node]->executed() && performance_data.has_output) {
+    		pqp_row_count = performance_data.output_row_count;
+  		}
+
+			// std::cout << "LQP Rowcount: " << lqp_row_count << std::endl;
+			// std::cout << "PQP Rowcount: " << pqp_row_count << std::endl;
```
Comment (leowe) [inline src/lib/visualization/cardinality_writer.cpp]: remove

File: src/lib/visualization/cardinality_writer.cpp [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+			// std::cout << _pqp_nodes[node]->name() << " " << _pqp_nodes[node]->description() << std::endl;
+			
+			const auto lqp_row_count = _cardinality_estimator.estimate_cardinality(_lqp_nodes[node]);
+			
+			auto pqp_row_count = NAN;
+			const auto& performance_data = *_pqp_nodes[node]->performance_data;
+
+  		if (_pqp_nodes[node]->executed() && performance_data.has_output) {
+    		pqp_row_count = performance_data.output_row_count;
+  		}
```
Comment (leowe) [inline src/lib/visualization/cardinality_writer.cpp]: indentation is all over the place --> run ./script/format.sh

File: src/lib/visualization/cardinality_writer.cpp [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+			// std::cout << "LQP Rowcount: " << lqp_row_count << std::endl;
+			// std::cout << "PQP Rowcount: " << pqp_row_count << std::endl;
+
+  		std::ofstream output_file;
+  		output_file.open ("./benchmark_cardinality_estimation.csv", std::ios_base::app);
+  		output_file << benchmark_item_id << "," << lqp_description << "," << pqp_description << "," << lqp_row_count << "," << pqp_row_count << "\n";
+  		output_file.close();
```
Comment (leowe) [inline src/lib/visualization/cardinality_writer.cpp]: Write Header somewhere? maybe print it to the command line? be careful: this code is executed multiple time

File: src/lib/visualization/cardinality_writer.cpp [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+		// Print debug information about possible different numbers of lqp and pqp nodes
+		std::ofstream output_file;
+		output_file.open ("./benchmark_cardinality_estimation_debug_info.csv", std::ios_base::app);
+		output_file << benchmark_item_id << "," << lqp_nodes_size << "," << pqp_nodes_size << "\n";
+		output_file.close();
+
+}
+
+void CardinalityWriter::_build_graph(const std::vector<std::shared_ptr<AbstractLQPNode>>& lqp_roots) {
+  std::unordered_set<std::shared_ptr<const AbstractLQPNode>> visualized_nodes;
```
Comment (leowe) [inline src/lib/visualization/cardinality_writer.cpp]: rename visualised_nodes and visualized_sub_queries

File: src/lib/visualization/cardinality_writer.cpp [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+                                   ExpressionUnorderedSet& visualized_sub_queries) {
+  // Avoid drawing dataflows/ops redundantly in diamond shaped Nodes
+  if (visualized_nodes.find(node) != visualized_nodes.end()) return;
+  visualized_nodes.insert(node);
+	_lqp_nodes.push_back(node);
+
+  auto node_label = node->description();
+  if (!node->comment.empty()) {
+    node_label += "\\n(" + node->comment + ")";
+  }
```
Comment (leowe) [inline src/lib/visualization/cardinality_writer.cpp]: do we need this here? I don't think so

File: src/lib/visualization/cardinality_writer.cpp [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+                                   std::unordered_set<std::shared_ptr<const AbstractOperator>>& visualized_ops) {
+  // Avoid drawing dataflows/ops redundantly in diamond shaped PQPs
+  if (visualized_ops.find(op) != visualized_ops.end()) return;
+  visualized_ops.insert(op);
+	_pqp_nodes.push_back(op);
+
+	auto op_label = op->name();
+  if (!op->description().empty()) {
+    op_label += "\\n(" + op->description() + ")";
+  }
```
Comment (leowe) [inline src/lib/visualization/cardinality_writer.cpp]: unnecessary

File: src/lib/visualization/cardinality_writer.cpp [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+      const auto limit = std::dynamic_pointer_cast<const Limit>(op);
+      _visualize_pqp_subqueries(op, limit->row_count_expression(), visualized_ops);
+    } break;
+
+    default: {
+    }  // OperatorType has no expressions
+  }
+}
+
+void CardinalityWriter::_visualize_pqp_subqueries(const std::shared_ptr<const AbstractOperator>& op,
```
Comment (leowe) [inline src/lib/visualization/cardinality_writer.cpp]: rename to something w/o visualize in its name

**Comment by phoeinx** [comment]
Now contained in #7

**Pull request closed without merging**